Provide note for 64bit integers

johanste opened the pull request:
JSON payloads containing numbers larger than JavaScripts Number.MAX_SAFE_INTEGER can cause silent truncation for JavaScript clients. Added blurb to suggest that services avoid this scenario.

johanste committed:
Provide note for 64bit integers

johanste commented:
FYI, @bterlson

bterlson commented on Guidelines.md:
```diff
@@ -1096,7 +1096,9 @@ If the delta link is no longer valid, the service MUST respond with _410 Gone_.
 ## 11. JSON standardizations
 ### 11.1. JSON formatting standardization for primitive types
-Primitive values MUST be serialized to JSON following the rules of [RFC4627][rfc-4627].
+Primitive values MUST be serialized to JSON following the rules of [RFC4627][rfc-4627]. 
```
4627 is obsoleted, update to https://tools.ietf.org/html/rfc8259?

bterlson commented on Guidelines.md:
```diff
@@ -1096,7 +1096,9 @@ If the delta link is no longer valid, the service MUST respond with _410 Gone_.
-Primitive values MUST be serialized to JSON following the rules of [RFC4627][rfc-4627].
+Primitive values MUST be serialized to JSON following the rules of [RFC4627][rfc-4627]. 
+
+**Important note for 64bit integers:** JavaScript will silently truncate integers larger than `Number.MAX_SIZE_INTEGER` (2^53-1). If the service is expected to return larger integer values, consider returning the value as a string. 
```
In JavaScript at least, it's Number.MAX_SAFE_INTEGER.

Worth noting that there is also Number.MIN_SAFE_INTEGER and numbers below -2^53+1 are not interoperable?

bterlson replied:
"Consider" is somewhat weaker than I was hoping 😄 We should at least "strongly consider" not exposing users to silent data loss I think!

johanste replied:
Strongly consider it is. And, apparently my JavaScript skills have deteriorated. Updated to MAX/MIN_SAFE...

johanste committed:
Address review feedback

Update outdated reference to RFC-4627 to refer to RFC-825.
Make recommendation to use string instead of number for large (and small) integers stronger.
Fix typo.

johanste committed:
Update Guidelines.md

johanste committed:
Fixed typo

bterlson commented on Guidelines.md:
```diff
@@ -1096,7 +1096,9 @@ If the delta link is no longer valid, the service MUST respond with _410 Gone_.
-Primitive values MUST be serialized to JSON following the rules of [RFC4627][rfc-4627].
+Primitive values MUST be serialized to JSON following the rules of [RFC8259][rfc-8259]. 
+
+**Important note for 64bit integers:** JavaScript will silently truncate integers larger than `Number.MAX_SAFE_INTEGER` (2^53-1) or numbers smaller than `Number.MIN_SAFE_INTEGER` (-2^53+1). If the service is expected to return integer values outside the range of safe values, strongly consider returning the value as a string in order to maximize interoperability adn avoid data loss.
```
s/adn/and

garethj-msft commented:
Great improvement - thanks.
